bug-fix(font-cache): Eliminated unwanted cloning of `cosmic_text::fontdb::Database` by passing it as a reference

namana-mecha opened the pull request:
# Objective
Fixes #70 

## Solution

`cosmic_test::fontdb::Database` was being cloned which was fixed by passing the `FontCache` itself as a mutable reference and making the changes accordingly.

### Before

```rust
    fn caches(&self) -> Caches {
        Caches {
            font: Arc::new(RwLock::new(FontCache::new(self.fonts.clone()))),
        }
    }
```
### After

```rust
    fn caches(&mut self) -> Caches {
        Caches {
            font: Arc::new(RwLock::new(&mut self.fonts_cache)),
        }
    }
```

## Testing

Flamegraph of the code after the fix

![image](https://github.com/user-attachments/assets/d395023f-7f15-4bbe-970e-60003b58821a)

Commit:
bug-fix(font-cache): Eliminated unwanted cloning of `cosmic_text::fontdb::Database` by passing it as a reference.

shoaibmerchant commented on core/src/ui.rs:
```diff
@@ -200,14 +200,14 @@ impl<
                         }
 
-                        let caches: crate::renderer::Caches = renderer.as_ref().unwrap().caches();
+                        let caches: crate::renderer::Caches = renderer.as_mut().unwrap().caches();
```
Do you need a mutable renderer for accessing caches?

namana-mecha replied:
Yes, because the ui updates the cache. That's why we were cloning in the first place or else a reference would have been fine.

```rust
let caches: crate::renderer::Caches = renderer.as_mut().unwrap().caches();
new.layout(&old, &mut caches.font.write().unwrap(), scale_factor);
do_render = new.render(caches, Some(&mut old), scale_factor);
```

the next line itself takes write access to the cache changes the cache if required and then performs the render itself.

shoaibmerchant replied:
Okay cool! Lets add a note explaining this in the code

shoaibmerchant commented:
Nice work @namana-mecha can you address the comment, this might cause a lock condition

Commit:
minor: Added comment to explain the mutable reference of `FontCache`

- The calculation of layout takes a mutable reference to the `FontCache`
in ui.rs. This commit will add a comment explaining the reasoning behind
it.
